### ISSUE TITLE
Applycal only reads "initial" ANTENA table and not "intermediate" from baseline selection

### ISSUE THREAD

**Issue by tikk3r** [lofar-astron/DP3#95]
Opening this here for reference with regards to https://github.com/lofar-astron/prefactor/issues/198.

A subselection of antennas is made using `msin.baseline`, to select only stations present in both the solutions and the MS they're applied to, but the subsequent applycal steps still read from the initial antenna table in the MS, causing a crash.

**Comment by aroffringa** [comment]
Thanks. So something to reproduce this, would be to have a selection like this:
`msin.baseline=[CR]S*&;!UK608HBA*;!DE602HBA*;!FR606HBA*;!DE604HBA*;!DE601HBA*;!DE605HBA*;!SE607HBA*;!DE603HBA*`
...and an `applycal` operation with a solution file that has only the selected stations?
Do you also specify `filter.remove=true` somewhere?

**Comment by tikk3r** [comment]
Yeah that should reproduce it. Here's the relevant bit in the parset:

https://github.com/lofar-astron/prefactor/blob/f45da871ea33fe0f3f7f2d5645c03a1b12453e2c/Pre-Facet-Target.parset#L191-L240

You say something interesting though. I noticed after your comment, that the filter part is commented out, but if I put that back in and pass `filter.baseline=[CR]S*&;!UK608HBA*;!DE602HBA*;!FR606HBA*;!DE604HBA*;!DE601HBA*;!DE605HBA*;!SE607HBA*;!DE603HBA*` then it seems to work. No immediate crash at least, as before.

Is there a slight difference in the way `msin.baseline` and `filter` work (`remove` aside)?

**Comment by aroffringa** [comment]
If you remove baselines in msin.baseline, the antenna table is not updated if you don't also have a `filter.remove=true` statement. I think it can be argued that it is supposed to work like this: if you don't 
filter.remove antennas, the antennas are not updated in the measurement, but remain existing. Only when removed, they truly disappear.

So bottom line, if the problem is solved with `filter.remove`, I think it works like intended, and you just have to add a filter.remove=true statement. The particular filter step doesn't have to do any selection itself, it is just there to remove unused antennas. Maybe @tammojan  can confirm this idea.

**Comment by tammojan** [comment]
The problem with the `filter=True` workaround is that it requires writing a new MS and reading that into DPPP, which is against the philosophy of DPPP.

The request could be implemented; some bookkeeping needs to be added. I've made bugs with this bookkeeping in the past, so I chose the 'easier' solution in ApplyCal.cc.

The code that would need to change is in OneApplyCal.cc::updateParms. There, a vector `parmValues` is made of size `info().antennaNames().size()`, this should be `info().antennaUsed().size()`. And all code below. This would cost half a day or so of work.

**Comment by tikk3r** [comment]
@aroffringa makes sense, the way you put it.

@tammojan In this case the fact that filter requires a new MS to be written is not that problematic I think, as the `ndppp_prep_target` step of prefactor is already writing new measurement sets anyway. It reads the data from the input directory, applies the calibrator solutions and outputs the new MS to the working directory, but maybe @adrabent can comment more on this, as he's working on it and I'm more of a user.

**Comment by adrabent** [comment]
@aroffringa
Maybe I am wrong, but if you unselect antennas with msin.baseline, e.g.,

`msin.baseline = [CR]S&*;!CS006HBA0*`

this antenna does not appear in the antenna list of the output measurement set. If you try to write to a CORRECTED_DATA column with that I am pretty sure DPPP will crash (what it actually should). I will try to reproduce that.

**Comment by adrabent** [comment]
I can reproduce the removal from the antenna table with `msin.baseline` easily. So it works as expected. At least I don't see the `!`-selected antenna anymore if I call `msoverview=test.MS verbose=T`

**Comment by adrabent** [comment]
@tikk3r
You are right. We can go back to `filter`, since we will always write new MS in such a case.

**Comment by aroffringa** [comment]
So am I summarizing this correctly if I say that this _is_ a bug in DP3, but @tikk3r can work in this case around it by using the filter? Than the bug can stay open, but won't have high priority.

**Comment by adrabent** [comment]
Indeed.
